File: 1 two-sum/Mister-Hope.cpp
/*
 * Runtime: 188 ms, faster than 36.71% of C++ online submissions for Two Sum.
 *
 * Memory Usage: 52.9 MB, less than 13.71% of C++ online submissions for Two
 * Sum.
 */

#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
 public:
  vector<int> twoSum(vector<int>& nums, int target) {
    unordered_map<int, int> prevMap;

    for (int i = 0; i < nums.size(); i++) {
      int j = target - nums[i];

      if (prevMap.count(j)) return {prevMap[j], i};
      prevMap[nums[i]] = i;
    }

    return {};
  }
};

File: 1 two-sum/Mister-Hope.ts
/*
 * Runtime: 124 ms, faster than 50.73% of TypeScript online submissions for Two Sum.
 *
 * Memory Usage: 48.7 MB, less than 5.74% of TypeScript online submissions for Two Sum.
 */

const twoSum = (nums: number[], target: number): number[] => {
  const arr = [];

  for (let i = 0; i < nums.length; i++) {
    let j = arr[target - nums[i]];
    if (j !== undefined) return [j, i];
    arr[nums[i]] = i;
  }

  throw new Error("No solution");
};
